### PR TITLE
Adding pin_memory kwarg to zeros, ones, empty, ... tensor constructors

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Make it possible to construct a pinned memory tensor without creating a storage first and without calling pin_memory() function. It is also faster, as copy operation is unnecessary.

Supported functions:
```python
torch.rand_like(t, pin_memory=True)
torch.randn_like(t, pin_memory=True)
torch.empty_like(t, pin_memory=True)
torch.full_like(t, 4, pin_memory=True)
torch.zeros_like(t, pin_memory=True)
torch.ones_like(t, pin_memory=True)
torch.tensor([10,11], pin_memory=True)
torch.randn(3, 5, pin_memory=True)
torch.rand(3, pin_memory=True)
torch.zeros(3, pin_memory=True)
torch.randperm(3, pin_memory=True)
torch.empty(6, pin_memory=True)
torch.ones(6, pin_memory=True)
torch.eye(6, pin_memory=True)
torch.arange(3, 5, pin_memory=True)
```

Part of the bigger: `Remove Storage` plan.

Now compatible with both torch scripts:
 `  _1 = torch.zeros([10], dtype=6, layout=0, device=torch.device("cpu"), pin_memory=False)`
and 
`  _1 = torch.zeros([10], dtype=6, layout=0, device=torch.device("cpu"))`

Same checked for all similar functions `rand_like`, `empty_like` and others

It is fixed version of #18455

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml]: This one isn't optional, but the other ones are. What's the difference?

Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml]: Doesn't the fact that these are not optional imply that you might break JIT scripts which mention any of the non-optional variants here?

File: aten/src/ATen/native_parse.py
Comment (ezyang) [inline aten/src/ATen/native_parse.py]: `matches = any(topt_representation[i] == topt for i, topt in enumerate(supported_topt))`

File: torch/onnx/symbolic.py
Comment (ezyang) [inline torch/onnx/symbolic.py]: That's some pretty bad boolean blindness

File: torch/onnx/symbolic.py
Comment (ezyang) [inline torch/onnx/symbolic.py]: I don't think you should pin the constant here. This tensor constant is going to be immediately serialized into the ONNX model. ONNX export will not treat a "pinned" tensor or an unpinned tensor any differently upon serialization. It's probably better to error in this case, and figure out what to do properly when a user complains about it.

File: tools/pyi/gen_pyi.py
Comment (ezyang) [inline tools/pyi/gen_pyi.py]: Don't use old style format strings. `"Error while processing function {}".format(fname)`

File: tools/autograd/templates/python_torch_functions.cpp
Comment (ezyang) [inline tools/autograd/templates/python_torch_functions.cpp]: Use of `ASSERT` here is inappropriate. Asserts are for error conditions that occur when we have a bug in our code. A user can easily trigger this condition; thus, an assert should not be used.

Comment (ezyang) [reply tools/autograd/templates/python_torch_functions.cpp]: Is there any particular reason this error check is done, as opposed to the actual implementation of `arange`? Error checks in the Python binding code don't show up in C++ API, and this seems like a reasonable thing to check for C++ too.

**Comment by ezyang** [comment]
Wasn't there going to be a test for TorchScript BC in this case?

**Review by ezyang — CHANGES_REQUESTED** [review]
see comments

File: aten/src/ATen/native/native_functions.yaml
Comment (VitalyFedyunin) [reply aten/src/ATen/native/native_functions.yaml]: I wish I could do it optional, however native function and jit generators only knows how to wrap all non-optional arguments to TensorOptions or all optional arguments to TensorOptions. There is no easy way to get combination of them (without major rewrite, which makes no sense as we migration off this generation anyway).

**Comment by VitalyFedyunin** [comment]
> Wasn't there going to be a test for TorchScript BC in this case?

Done manually with binary files.

File: tools/autograd/templates/python_torch_functions.cpp
Comment (VitalyFedyunin) [reply tools/autograd/templates/python_torch_functions.cpp]: We also call check_out_type_matches here. And it does exactly same, pre-validation of passed parameters and throwing AT_ERROR. 

There is no need to implement this check inside of cpp function as in case of output tensor it doesn't take options.

`wrap(dispatch_arange(r.scalar(0), r.scalar(1), r.scalar(2), r.tensor(3)).set_requires_grad(r.toBool(8)))`

where 
 0 - start
 1 - end
 2 - step 
 3 - output
 8 - requires_grad

File: torch/onnx/symbolic.py
Comment (VitalyFedyunin) [reply torch/onnx/symbolic.py]: Unfortunately all args except ScalarType are ignored in this function

File: torch/onnx/symbolic.py
Comment (VitalyFedyunin) [reply torch/onnx/symbolic.py]: Maybe warning?

File: aten/src/ATen/native/native_functions.yaml
Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml]: I think my real question is, why weren't *all* of these optional in the first place?

File: torch/onnx/symbolic.py
Comment (ezyang) [reply torch/onnx/symbolic.py]: I find warnings are not very actionable. Either it is sound to export pinned tensors as normal ones (because, e.g., nothing in ONNX distinguishes between pinned and unpinned--a claim I find plausible), or it's not, in which case we should error.

File: tools/autograd/templates/python_torch_functions.cpp
Comment (ezyang) [reply tools/autograd/templates/python_torch_functions.cpp]: OK, if there's no C++ out + TensorOptions variant, that's sufficient for me.

**Comment by ezyang** [comment]
> Done manually with binary files.

The general expectation is that if you did manual testing to verify if a change worked, you should describe in the PR how exactly you tested it ;) (That's what "Test Plan" in FB infrastructure is all about). But I think there's a case to be made for actually have a real test in the test suite for this case, so that we can avoid breaking it in the future. What ended up being difficult about constructing an ad hoc JIT IR with the missing kwarg field for this?

File: tools/autograd/templates/python_torch_functions.cpp
Comment (ezyang) [reply tools/autograd/templates/python_torch_functions.cpp]: > We also call check_out_type_matches here. And it does exactly same, pre-validation of passed parameters and throwing AT_ERROR.

Just double-checking; this comment is in reference to "Use of ASSERT here is inappropriate. Asserts are for error conditions that occur when we have a bug in our code. A user can easily trigger this condition; thus, an assert should not be used." is that right?

Comment (ezyang) [reply tools/autograd/templates/python_torch_functions.cpp]: I guess it wasn't, because this is still broken:

```
>>> torch.arange(2, out=x, pin_memory=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: !r.toBool(5) ASSERT FAILED at ../torch/csrc/autograd/generated/python_torch_functions.cpp:119, please report a bug to PyTorch.  `pin_memory` and `out` parameters are incompatible
```

**Comment by VitalyFedyunin** [comment]
> What ended up being difficult about constructing an ad hoc JIT IR with the missing kwarg field for this?

This type of the test will cover only this particular tiny BC (much more easier to test manually). Making proper test will require or storing binary, or inlining all zipped files inside of test code (including attributes.pkl binary)

File: aten/src/ATen/native/native_functions.yaml
Comment (VitalyFedyunin) [reply aten/src/ATen/native/native_functions.yaml]: This is part of the bigger `what we do with tensor options` discussion.

**Comment by ezyang** [comment]
> This type of the test will cover only this particular tiny BC (much more easier to test manually)

I mean, sure, it will only cover this particular BC, but the point of adding regression tests when a regression happens is that we have some updated priors that this particular aspect of the system is likely to break. And it doesn't seem unreasonable to me that it will break exactly the same way the next time we need to add another option to `TensorOptions`.

File: aten/src/ATen/native/native_functions.yaml
Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml]: > Doesn't the fact that these are not optional imply that you might break JIT scripts which mention any of the non-optional variants here?

Answering my question, no, JIT scripts should not be broken, because all occurrences of `pin_memory` are given default arguments.

I think there should still be a Note talking about the inconsistency of `pin_memory` in these sites. Put it in `native_functions.yaml` (or maybe the README in native)

File: tools/autograd/templates/python_torch_functions.cpp
Comment (VitalyFedyunin) [reply tools/autograd/templates/python_torch_functions.cpp]: What I'm trying to say is that 

```
arange(Scalar end, *, Tensor out=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)
arange(Scalar start, Scalar end, Scalar step=1, *, Tensor out=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)
```
Is incorrect approach (if we are trying to avoid AT_ERROR), with or without `pin_memory` introduced.

The proper approach is to have 4 signatures:

```
arange(Scalar end, *, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)
arange(Scalar start, Scalar end, Scalar step=1, *, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)

arange(Scalar end, *, Tensor out, bool requires_grad=False)
arange(Scalar start, Scalar end, Scalar step=1, *, Tensor out, bool requires_grad=False)
```

In this case `check_out_type_matches` will go away as well as `AT_ASSERTM(!r.toBool(5),`. But this requires bigger review of all function and out of scope.

Comment (ezyang) [reply tools/autograd/templates/python_torch_functions.cpp]: Erm, rewinding a sec: my original ask was that you use `AT_CHECK` rather than `AT_ASSERT`. This is why I posted a snippet of a simple user code snippet triggering the assert. **The assert literally tells the user to file a bug** and we don't want that in this situation. I'm not asking for any restructuring of the code in question. In particular, I'm not asking you to completely remove the error check.

I agree your four signatures is better, and I also agree we shouldn't do it in this PR.

File: aten/src/ATen/native/native_functions.yaml
Comment (gchanan) [reply aten/src/ATen/native/native_functions.yaml]: There's a lot of discussion about this function when it doesn't even matter (outside of not breaking JIT) -- no one calls this function directly.  So presumably we can just do whatever makes our job easier here (giving everything a default, not using TensorOptions, etc. as long as it doesn't break JIT).

File: torch/csrc/utils/tensor_new.cpp
Comment (gchanan) [inline torch/csrc/utils/tensor_new.cpp]: what's the story here?  Don't the parameter indices need to be updated here (and why don't the rest of the functions below have the parameter?).

File: aten/src/ATen/native/native_functions.yaml
Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml]: That's my fault, I stuck the PR comment on the very first occurrence that had this problem. If you look at the rest of `native_functions.yaml` there are a bunch of other, more public, functions that look this way too. (In the end it's a moot point, I don't believe there is any BC-breakage here, regardless of the status of the function.)

**Comment by VitalyFedyunin** [comment]
> > This type of the test will cover only this particular tiny BC (much more easier to test manually)
> 
> I mean, sure, it will only cover this particular BC, but the point of adding regression tests when a regression happens is that we have some updated priors that this particular aspect of the system is likely to break. And it doesn't seem unreasonable to me that it will break exactly the same way the next time we need to add another option to `TensorOptions`.

Will land #19174 first (need stamp ;) )

**Comment by ezyang** [comment]
You don't have to wait for #19174 to land first before landing this one; your assurance is good enough for me :)

File: torch/csrc/utils/tensor_new.cpp
Comment (VitalyFedyunin) [reply torch/csrc/utils/tensor_new.cpp]: Nice catch! I think we already had discussion about named parameters in PythonArgParser, right?

**Review by ezyang — APPROVED** [review]
I didn't heavily rereview the Python arg parser code, let me know if you want a careful audit.

**Pull request closed without merging**
